video/out/vulkan/context_wayland.c

```diff
@@ -15,17 +15,78 @@
  * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <poll.h>
+
+#include "osdep/timer.h"
 #include "video/out/gpu/context.h"
 #include "video/out/wayland_common.h"
-
 #include "common.h"
 #include "context.h"
 #include "utils.h"
 
 struct priv {
     struct mpvk_ctx vk;
+    bool callback_init;
+    bool callback_wait;
 };
 
+static const struct wl_callback_listener frame_listener;
+
+static void frame_callback(void *data, struct wl_callback *callback, uint32_t time)
+{
+    struct ra_ctx *ctx = data;
+    struct priv *p = ctx->priv;
+    struct vo_wayland_state *wl = ctx->vo->wl;
+
+    if (callback)
+        wl_callback_destroy(callback);
+
+    wl->frame_callback = wl_surface_frame(wl->surface);
+    wl_callback_add_listener(wl->frame_callback, &frame_listener, ctx);
+    p->callback_wait = false;
+}
+
+static const struct wl_callback_listener frame_listener = {
+    frame_callback,
+};
+
+static void wayland_vk_swap_buffers(struct ra_ctx *ctx)
+{
+    struct priv *p = ctx->priv;
+    struct vo_wayland_state *wl = ctx->vo->wl;
+
+    struct pollfd fds[2] = {
+        {.fd = wl->display_fd,     .events = POLLIN },
+        {.fd = wl->wakeup_pipe[2], .events = POLLIN },
+    };
+
+    wl_display_dispatch_pending(wl->display);
+
+    double vblank_time = (1/wl->current_output->refresh_rate)*1e6;
+    int64_t finish_time = mp_time_us() + vblank_time;
+
+    while (p->callback_wait) {
+        if (finish_time - mp_time_us() <= 0)
+            break;
+
+        wl_display_roundtrip(wl->display);
+
+        // Divide the difference by an extra factor of 2 to get
+        // the halfway point between now and the finish time.
+        int64_t poll_time = (finish_time - mp_time_us())/2000;
+        poll(fds, 2, poll_time);
+    }
```

Comment:
This loop is duplicated from the EGL one. The EGL one might even count as duplicated with the "normal" even loop. Does wl_display_roundtrip dispatch user callbacks or so?

Reply:
`wl_display_roundtrip` blocks pending requests so they can be processed by the server. It's needed so the timings of reading the events are correct. I don't think this counts as a duplicate of the normal event loop. mpv is written with the assumption that buffer swap blocks for vsync, but the wayland buffer swap has the nasty behavior of indefinitely blocking. So some extra logic is need for it.